File: app/src/main/java/com/example/myaac/data/nlp/MorphologyService.kt
package com.example.myaac.data.nlp

import simplenlg.features.Feature
import simplenlg.features.Tense
import simplenlg.framework.LexicalCategory
import simplenlg.framework.NLGFactory
import simplenlg.lexicon.Lexicon
import simplenlg.phrasespec.SPhraseSpec
import simplenlg.realiser.english.Realiser

class MorphologyService {

    private val lexicon: Lexicon = Lexicon.getDefaultLexicon()
    private val nlgFactory: NLGFactory = NLGFactory(lexicon)
    private val realiser: Realiser = Realiser(lexicon)

    /**
     * generating variations for a given word and part of speech.
     * For example, "eat" (VERB) -> ["ate", "eating", "will eat", "eats"]
     */
    fun getVariations(word: String, category: String): List<String> {
        val variations = mutableListOf<String>()
        val lexicalCategory = when (category.uppercase()) {
            "VERB" -> LexicalCategory.VERB
            "NOUN" -> LexicalCategory.NOUN
            "ADJECTIVE" -> LexicalCategory.ADJECTIVE
            else -> return emptyList()
        }

        try {
            when (lexicalCategory) {
                LexicalCategory.VERB -> {
                    // Base form
                    variations.add(word) 
                    
                    // Past
                    val pastClause: SPhraseSpec = nlgFactory.createClause()
                    pastClause.setVerb(word)
                    pastClause.setFeature(Feature.TENSE, Tense.PAST)
                    variations.add(realiser.realiseSentence(pastClause).trim().trimEnd('.'))

                    // Present Progressive (ing)
                    val progressiveClause: SPhraseSpec = nlgFactory.createClause()
                    progressiveClause.setVerb(word)
                    progressiveClause.setFeature(Feature.PROGRESSIVE, true)
                    progressiveClause.setFeature(Feature.TENSE, Tense.PRESENT)
                     // SimpleNLG makes full sentences, we might get "is eating". 
                     // Ideally we want just the morphological form "eating".
                     // SimpleNLG is a bit sentence-heavy. 
                     // Let's try to just get the word form if possible, 
                     // or clean up the "is" if it adds auxiliary.
                     // Actually, for AAC "is eating" is often what we want for "present progressive".
                    variations.add(realiser.realiseSentence(progressiveClause).trim().trimEnd('.'))

                    // Future
                    val futureClause: SPhraseSpec = nlgFactory.createClause()
                    futureClause.setVerb(word)
                    futureClause.setFeature(Feature.TENSE, Tense.FUTURE)
                    variations.add(realiser.realiseSentence(futureClause).trim().trimEnd('.'))
                }
                LexicalCategory.NOUN -> {
                     // Singular
                     variations.add(word)

                     // Plural
                     val nounPhrase = nlgFactory.createNounPhrase(word)
                     nounPhrase.setPlural(true)
                     variations.add(realiser.realise(nounPhrase).realisation.trim())
                }
                else -> {
                    // Add other categories if needed
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return listOf(word) // Fallback
        }

        return variations.distinct()
    }
}

File: web/src/services/morphologyService.ts
// Replacement for data/nlp/MorphologyService.kt, which wrapped SimpleNLG (a JVM-only NLG
// library with no JS equivalent).
//
// Per the migration plan's "Replacing SimpleNLG" section: MorphologyService.getVariations() had
// exactly one call site (MainActivity.kt:980-982 — the "grammar popup" word-variation picker
// dialog reached via a button's long-press/grammar action) and, per the codebase, no Hebrew
// lexicon actually wired up in practice (English-only despite the app being bilingual). Rather
// than port or replace SimpleNLG, this issues a single Gemini prompt asking for the word's
// morphological variations directly in the target language — a strict capability upgrade for
// Hebrew (Gemini vs. no working Hebrew lexicon today) — falling back to a small hardcoded
// English +ed/+ing/+s suffix-rule table when Gemini is unreachable (offline, or the call fails),
// matching the app's existing "AI call with a static-rule fallback" idiom used elsewhere
// (see nlp/hybridPredictionEngine.ts).
//
// Call-site contract (from MainActivity.kt:980-982, preserved here):
//   const verbs = await morphologyService.getVariations(word, 'VERB');
//   const nouns = await morphologyService.getVariations(word, 'NOUN');
//   const adjectives = await morphologyService.getVariations(word, 'ADJECTIVE');
//   const combined = [...new Set([...verbs, ...nouns, ...adjectives])];
//   // shown only if combined.length > 0 && combined is not just [word]

import { generateGeminiText } from './geminiService';

export type MorphologyCategory = 'VERB' | 'NOUN' | 'ADJECTIVE';

export class MorphologyService {
  /**
   * Generate variations for a given word and part of speech, e.g. "eat" (VERB) ->
   * ["eat", "ate", "eating", "will eat"], or "cat" (NOUN) -> ["cat", "cats"].
   * Always includes the base word itself, mirroring the Kotlin source's behavior of adding the
   * base form as the first variation.
   */
  async getVariations(
    word: string,
    category: MorphologyCategory,
    languageCode: string = 'en',
  ): Promise<string[]> {
    const trimmed = word.trim();
    if (!trimmed) return [];

    try {
      const aiVariations = await this.getVariationsFromGemini(trimmed, category, languageCode);
      if (aiVariations.length > 0) {
        return dedupe([trimmed, ...aiVariations]);
      }
    } catch (e) {
      console.warn('MorphologyService: Gemini call failed, falling back to static rules', e);
    }

    // Offline / Gemini-unreachable fallback. English-only, matching the plan's scope for this
    // fallback path.
    return dedupe([trimmed, ...englishSuffixFallback(trimmed, category)]);
  }

  private async getVariationsFromGemini(
    word: string,
    category: MorphologyCategory,
    languageCode: string,
  ): Promise<string[]> {
    const languageName = languageCode === 'iw' || languageCode === 'he' ? 'Hebrew' : 'English';
    const formsClause =
      category === 'VERB'
        ? 'its base form, past tense, present progressive ("-ing" / continuous) form, and future tense form'
        : category === 'NOUN'
          ? 'its singular and plural forms'
          : 'its base form and common comparative/intensified forms, if applicable';

    const prompt = `Given this word and part of speech, return its base/past/progressive/future or singular/plural forms in ${languageName}.

Word: "${word}"
Part of speech: ${category}

Return ${formsClause}.
Return ONLY the forms, separated by commas. No explanations, no numbering.`;

    const text = await generateGeminiText(prompt);
    if (text == null) return [];
    return text
      .split(',')
      .map((s) => s.trim())
      .filter((s) => s.length > 0);
  }
}

function dedupe(words: string[]): string[] {
  return [...new Set(words)];
}

// ===== English suffix-rule fallback (+ed / +ing / +s), for offline use of the grammar popup =====

const VOWELS = new Set(['a', 'e', 'i', 'o', 'u']);

function isConsonant(ch: string): boolean {
  return /[a-z]/i.test(ch) && !VOWELS.has(ch.toLowerCase());
}

/** Consonant-Vowel-Consonant check for the final 3 letters (double-the-final-consonant rule). */
function endsInCvc(word: string): boolean {
  if (word.length < 3) return false;
  const [c1, v, c2] = word.slice(-3).toLowerCase();
  return isConsonant(c1) && VOWELS.has(v) && isConsonant(c2) && c2 !== 'w' && c2 !== 'x' && c2 !== 'y';
}

function endsInConsonantY(word: string): boolean {
  return word.length >= 2 && word.toLowerCase().endsWith('y') && isConsonant(word[word.length - 2]);
}

function pluralOrThirdPersonS(word: string): string {
  const lower = word.toLowerCase();
  if (/(s|x|z|ch|sh)$/.test(lower)) return `${word}es`;
  if (endsInConsonantY(word)) return `${word.slice(0, -1)}ies`;
  return `${word}s`;
}

function pastTenseEd(word: string): string {
  const lower = word.toLowerCase();
  if (lower.endsWith('e')) return `${word}d`;
  if (endsInConsonantY(word)) return `${word.slice(0, -1)}ied`;
  if (endsInCvc(word)) return `${word}${word[word.length - 1]}ed`;
  return `${word}ed`;
}

function presentParticipleIng(word: string): string {
  const lower = word.toLowerCase();
  if (lower.endsWith('ie')) return `${word.slice(0, -2)}ying`;
  if (lower.endsWith('e') && !lower.endsWith('ee') && !lower.endsWith('oe')) {
    return `${word.slice(0, -1)}ing`;
  }
  if (endsInCvc(word)) return `${word}${word[word.length - 1]}ing`;
  return `${word}ing`;
}

function englishSuffixFallback(word: string, category: MorphologyCategory): string[] {
  switch (category) {
    case 'VERB':
      return [pastTenseEd(word), presentParticipleIng(word), `will ${word}`];
    case 'NOUN':
      return [pluralOrThirdPersonS(word)];
    case 'ADJECTIVE':
      return [];
  }
}
